Add function for reverse engineering savename

sebastianpech opened the pull request:
Implements a function to parse the savename into a dict and extract the pre- and suffix (#38).
I thought of some special cases
- prefix is a path containing a folder named something like "a=1.23_mode=double"
- key or value contains `connector` (no precise solution, I'm adding them to the value part)
- last number is a float so the "." doesn't denote the start of the suffix

Datseris commented on test/naming_tests.jl:
```diff
@@ -57,3 +57,75 @@ s = savename(e; allowedtypes = (Any,), expand = ["c"])
+function test_convert(prefix::AbstractString,c;kwargs...)
+    name = savename(prefix,c;kwargs...)
+    prefix, _b, suffix = DrWatson.parse_savename(name)
+    dicts_equal(_b,c) && prefix == prefix && suffix == ""
```
Hm, maybe `prefix == prefix ` here is a typo? (This can only be `true`, I would assume that you wanted to name the prefix of the previous line as something like `prefix2`

sebastianpech replied:
Oh yes, that's nonsense :) I'll fix it

Datseris commented on test/naming_tests.jl:
```diff
@@ -57,3 +57,75 @@ s = savename(e; allowedtypes = (Any,), expand = ["c"])
+function test_convert(c,suffix::AbstractString;kwargs...)
+    name = savename(c,suffix;kwargs...)
+    prefix, _b, suffix = DrWatson.parse_savename(name)
+    dicts_equal(_b,c) && suffix == suffix && prefix == ""
```
same comment here for `suffix`.

Maybe I have misunderstood something, so please correct me if need be.

Datseris commented on src/naming.jl:
```diff
@@ -250,3 +250,89 @@ end
+"""
+    parse_savename(filename::AbstractString; kwargs...)
+                        parsetypes = (Int, Float64),
+                        connector = "_")
```
I assume that the extra keywords here were forgotten in the call signature, since you very nicely explain them later on in their own section.

sebastianpech replied:
Oh yes. I usually have them in the call signature, but wanted to match your style

Datseris commented on src/naming.jl:
```diff
@@ -250,3 +250,89 @@ end
+        # if no dot is after the last "="
+        # there is no suffix
+        name, suffix = savename_part,""
+    else
```
Wow, this is very impressive; you have implemented a very smart function, congrats! :)

Datseris commented on src/naming.jl:
```diff
@@ -250,3 +250,89 @@ end
+    # This way it is possible to corretly match something like
+    # string_with_underscore_a=10.1 as the next match for the regex is
+    # string_with[_underscore_a=]10.1
+    name_seperator = Regex("$connector[^$connector]+=")
```
Again, this is excellent; I would never have been able to do this myself, since I don't really know much about Regex... Jonas knows much more, I assume he would be even happier than me to see this!

sebastianpech commented:
Thanks for the nice package. I've attempted to implement something similar many times in the recent year, but was never fully satisfied.
DrWatson doesn't yet cover all the stuff I need (or use) in my workflow but I'm sure we manage to create a  nice package as a community.

sebastianpech commented:
I changed my mind about parsing the connector symbol. In my opinion it's more likely (especially for the underscore) to have it in the variable name instead of the variable value. So I turned the regex around and now the value is delimited by = and the closest connector symbol. So stuff like `l_ac=10.1` is now correctly parsed but `some_string=string_with_underscore_l_ac=10.1` results in `some_string = "string"` and `with_underscore_l_ac=10.1`. I also added a note to the docstring about the parsing behavior.

sebastianpech commented:
Should I merge the current master here?

JonasIsensee commented:
Hey, good work.

Just a quick comment.
if you really do want to allow for values with _
You could probably still escape them in the savename
function.
`some_string='my_string'_next_var=5`

But I'm not sure sure if that this would be a good idea..
(im also not sure if those are valid characters for filenames in Windows)

sebastianpech commented:
Thanks! Yes escaping would be an option, but as you said, I also don't think it's a very good idea to put `'` into the filename or actually in general to put too much data into the filename.

sebastianpech commented:
One thing just came to my mind on the way to work. I'm not taking into account that `connector` can be an arbitrary string. Do you want to keep allowing that or should we restrict it to beeing a `Char`?

Datseris commented:
> Should I merge the current master here?

Yes please.

> I'm not taking into account that connector can be an arbitrary string. Do you want to keep allowing that or should we restrict it to being a Char?

I don't think this is worth the effort to expand to multi-string connectors. In my eyes `parse_savename` should target the basic outcomes of `savename`, not complicated ones. I find it fair to leave it as is, and you can throw an error in multi-string connectors.

> You could probably still escape them in the savename function.
> `some_string='my_string'_next_var=5`
> But I'm not sure sure if that this would be a good idea..

I agree, I don't think this is a good plan either.
